rfcs/text/0013_background_sessions.md

```diff
@@ -0,0 +1,484 @@
+
+### New Session Service
+
+We will expose a new frontend `session` service on the `data` plugin `search` service. 
+
+The service will expose the following APIs:
+
+```ts
+interface ISessionService {
+   /** 
+    * Returns the current session ID
+    */
+   getId: () => string;
+
+   /**
+    * Start a new session, by generating a new session ID.
+    */
+   start: () => string;
+
+   /**
+    * Restore a session. Any searches executed using this ID will attempt to reload saved data.
+    * A restored session may be expired, meaning that following search requests will fail. 
+    * @param sessionId Session ID to restore.`.
+    */
+   restore: (sessionId: string) => void;
+
+   /**
+    * Close the current session and start a new one
+    */
+   clear: () => void; 
+
+   /**
+    * Store a session.
+    * @param sessionId Session ID to store. Probably retrieved from `sessionService.get()`.
+    * @param name A display name for the session.
+    * @param url TODO: is the URL provided here? How?
+    * @param searchIdMapping An optional mapping of hashed requests mapped to the corresponding searchId.
+    * @returns The stored `BackgroundSessionAttributes` object
+    * @throws Throws an error in OSS.
+    */
+   store: (sessionId: string, name: string, url: string, searchIdMapping?: Record<string, string>)) => Promise<BackgroundSessionAttributes>
+
+   /**
+    * @returns Is the current session stored (i.e. is there a saved object correspondng with this sessionId).
+    */
+   isStored: () => boolean;
+
+   /** 
+    * @returns Is the current session restored (i.e. was loaded by calling `sessionService.restore()`).
+    */
+   isRestore: () => boolean;
+
+   /**
+    * Mark a session as and all associated searchIds as expired.
+    * Cancels active requests, if there are any.
+    * @param sessionId Session ID to store. Probably retrieved from `sessionService.get()`.
+    * @returns success status
+    * @throws Throws an error in OSS.
+    */
+   expire: (sessionId: string) => Promise<boolean> 
+   
+   /**
+    * Extend a session and all associated searchIds.
+    * @param sessionId Session ID to extend. Probably retrieved from `sessionService.get()`.
+    * @param extendBy Time to extend by, can be a relative or absolute string.
+    * @returns success status
+    * @throws Throws an error in OSS.
+    */
+   extend: (sessionId: string, extendBy: string)=> Promise<boolean>
+
+   /**
+    * @param TODO: input arguments
+    * @returns a filtered list of BackgroundSessionAttributes objects.
+    * @throws Throws an error in OSS.
+    */
+   list: (options: SavedObjectsFindOptions) => BackgroundSessionAttributes[]
+}
+```
+
+### Search Service Changes
```

Comment:
Yes, in fact @lizozom and I have talked about removing the `SearchInterceptor` entirely since it's just an implementation detail of the search service. We may want to just move these methods directly into the search service.

Reply:
++